scripts/Muon/GUI/ElementalAnalysis2/grouping_widget/ea_grouping_table_widget_view.py

```diff
@@ -0,0 +1,328 @@
+
+        self.remove_group_action = self._context_menu_remove_group_action(self.remove_group_button.clicked.emit)
+
+        if self._disabled:
+            self.remove_group_action.setEnabled(False)
+
+        self.menu.addAction(self.remove_group_action)
+
+        self.menu.popup(QtGui.QCursor.pos())
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # Context menu on right-click in the table
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def _context_menu_remove_group_action(self, slot):
+        if len(self._get_selected_row_indices()) > 1:
+            # use plural if >1 item selected
+            remove_group_action = QtWidgets.QAction('Remove Groups', self)
+        else:
+            remove_group_action = QtWidgets.QAction('Remove Group', self)
+        if self.num_rows() == 0:
+            remove_group_action.setEnabled(False)
+        remove_group_action.triggered.connect(slot)
+        return remove_group_action
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # Slot connections
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def on_user_changes_group_name(self, slot):
+        self._validate_group_name_entry = slot
+
+    def on_remove_group_button_clicked(self, slot):
+        self.remove_group_button.clicked.connect(slot)
+
+    def on_table_data_changed(self, slot):
+        self._on_table_data_changed = slot
+
+    def on_cell_changed(self, _row, _col):
+        if not self._updating:
+            self._on_table_data_changed(_row, _col)
+
+    # ------------------------------------------------------------------------------------------------------------------
+    #
+    # ------------------------------------------------------------------------------------------------------------------
+
+    def get_table_item_text(self, row, col):
+        return self.grouping_table.item(row, col).text()
+
+    def get_table_item(self, row, col):
+        return self.grouping_table.item(row, col)
+
+    def set_to_analyse_state(self, row, state):
+        checked_state = QtCore.Qt.Checked if state is True else QtCore.Qt.Unchecked
+        item = self.get_table_item(row, inverse_group_table_columns['to_analyse'])
+        item.setCheckState(checked_state)
+
+    def set_to_analyse_state_quietly(self, row, state):
+        checked_state = QtCore.Qt.Checked if state is True else QtCore.Qt.Unchecked
+        item = self.get_table_item(row, inverse_group_table_columns['to_analyse'])
+        self.grouping_table.blockSignals(True)
+        item.setCheckState(checked_state)
+        self.grouping_table.blockSignals(False)
+
+    def get_table_contents(self):
+        if self._updating:
+            return []
+        ret = []
```

Comment:
what is ret?

Reply:
This method was copied over from the Muon table widget view as one of the necessary methods to make it all work. No idea what ret is!

Reply:
I think its "retun_value". A better variable name would be: `table_by_row`, each index contains a list of what is in each row